### PR TITLE
✨🖍🐛 Lightbox captions changes.

### PR DESCRIPTION
**Pull request by sparhami**
- Refactor lightbox captions into a separate file.
- Fix downwards movement at the start of the captions expanding by
having consistent padding both before and after the expansion.
- Allow opening lightbox with the captions already expanded via an
action.
- Apply captions fading mask when switching slides, which is less
jarring when the new slide's captions are overflowing.
- Keep the close and gallery view buttons visible when the caption is
expanded. This is important when opening when captions already expanded.
  * Added a fade out effect to the top of the text, so it does not clash
  with the buttons.
- Change the caption fade out mask to use `mask-image` instead of a
gradient. This works well when the captions are overlaying the image
(e.g. if you enter pan-zoom mode or the image is in portrait).
  * Add a text-shadow around the captions as a replacement, so when the
  captions are on top of an image, they are still readable. This also
  improves the contrast when the captions are expanded.
- Fix stacking order bug when fading in.

The existing version of the lightbox (without opening with expanded captions) can be tested here:
https://output.jsbin.com/vebebuy

The change can be tested here:
https://amp-polish.firebaseapp.com/manual/amp-lightbox-gallery-launch-captions.amp.html

### CODE REVIEW

**Comment by sparhami** [comment]
A few things to compare on the above links:

- Expand the images into pan-zoom mode by double tapping, and comparing how the captions look.
- The expanded caption behavior (especially the second image, which scrolls)
- The caption expansion animation

File: extensions/amp-lightbox-gallery/0.1/amp-lightbox-gallery.js [hunk 1]
```diff
@@ -1062,35 +991,32 @@ export class AmpLightboxGallery extends AMP.BaseElement {
...
-      userAssert(target,
-          'amp-lightbox-gallery.open: element with id: %s not found', targetId);
-    }
-    this.openLightboxGallery_(dev().assertElement(target)).then(() => {
-      return this.history_.push(this.close_.bind(this));
-    }).then(historyId => {
-      this.historyId_ = historyId;
-    });
+  handleOpenAction_(invocation) {
+    const args = invocation.args || {};
```
Comment (alanorozco) [inline extensions/amp-lightbox-gallery/0.1/amp-lightbox-gallery.js]: Nit for terseness. Also, personal preference nit: `handle` is redundant imo, just `openOnAction_`?
```js
openOnAction_({args = {}}) {
  const id = args['id'];
  // ...
```

File: extensions/amp-lightbox-gallery/0.1/amp-lightbox-gallery.js [hunk 2]
```diff
@@ -1110,14 +1036,16 @@ export class AmpLightboxGallery extends AMP.BaseElement {
...
+          'slideChange', this.boundSlideChangeHandler_);
 
       this.setupGestures_();
       this.setupEventListeners_();
 
       return this.carousel_.signals().whenSignal(CommonSignals.LOAD_END);
-    }).then(() => this.openLightboxForElement_(element))
+    })
+        .then(() => {
+          return this.openLightboxForElement_(element, expandDescription);
```
Comment (alanorozco) [inline extensions/amp-lightbox-gallery/0.1/amp-lightbox-gallery.js]: Any reason for expanding the short return syntax? I find that the short syntax can diminish the temptation to make the callback more complex and instead encapsulate behavior in the callee.

Either way, just a personal preference thing.

File: extensions/amp-lightbox-gallery/0.1/amp-lightbox-gallery.js [hunk 3]
```diff
@@ -1367,7 +1297,13 @@ export class AmpLightboxGallery extends AMP.BaseElement {
         toggle(this.element, true);
       }
 
+
       setStyles(this.element, {
+        // These are needed to stack correctly given that we give the element
```
Comment (alanorozco) [inline extensions/amp-lightbox-gallery/0.1/amp-lightbox-gallery.js]: So this is less than ideal right now. With `auto-lightbox` many fixed headers are indexed higher, which causes rendering issues. Not an issue for now but we need to keep it in mind, not sure if the `FixedLayer` hacks actually break anything.

Also moving this here has the side effect of these not being appended to the z-index file.

I'd say just create a fake (obviously namespaced) selector in the CSS for this extension, just for documenting, and regenerate `css/Z_INDEX.md` (`gulp get-zindex`).

File: extensions/amp-lightbox-gallery/0.1/lightbox-caption.js [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import {htmlFor} from '../../../src/static-template';
+
+/**
+ * @enum {string}
+ */
+export const OverflowState = {
```
Comment (alanorozco) [inline extensions/amp-lightbox-gallery/0.1/lightbox-caption.js]: 👍

File: extensions/amp-lightbox-gallery/0.1/amp-lightbox-gallery.js (hunk 3 above)
Comment (alanorozco) [reply extensions/amp-lightbox-gallery/0.1/amp-lightbox-gallery.js]: (Also, personal TODO: there's no reason why these shouldn't be able to be parsed from the get z index generation, it should be trivial to do.)

Comment (sparhami) [reply extensions/amp-lightbox-gallery/0.1/amp-lightbox-gallery.js]: Moved most of the styles into CSS and refactored the controls fade out logic.

**Pull request merged**